src/System.Linq/src/System/Linq/DefaultIfEmpty.cs

```diff
@@ -103,6 +103,20 @@ public List<TSource> ToList()
                 return list;
             }
 
+            public HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                List<TSource> list = _source.ToList();
+
+                if (list.Count == 0)
+                {
+                    list.Add(_default);
+                }
```

Comment:
Why go through creating a list here? The only way a hashset created from the source could be zero-size is if the source was, and likewise for non-zero:

```C#
HashSet<TSource> hashSet = new HashSet<TSource>(list, comparer);
if (hashSet.Count == 0)
{
  hashSet.Add(_default);
}